KIPs/kip-113.md

```diff
@@ -0,0 +1,216 @@
+- Proof-of-possession (pop): the signature of the hash of the public key (96 bytes)
+
+### Smart Contracts Overview
+
+The proposed smart contract will be implemented in Solidity and will be compatible with the Ethereum Virtual Machine (EVM).
+
+The smart contract will have the following features:
+
+- Register a public key and a proof-of-possession
+- Unregister a public key
+- View the registered public keys
+
+#### Methods
+
+##### registerPublicKey
+
+Registers the given public key associated with the `msg.sender` address.
+
+The function validates the following requirements:
+
+- The function MUST revert if `publicKey.length != 48` or `pop.length != 96`
+- The function MUST revert if `publicKey` or `pop` is equal to zero
+- The function SHOULD authenticate and authorize `msg.sender`
+
+The function emits a `PubkeyRegistered` event.
+
+_Note_ The function is not able to verify the validity of the public key and the proof-of-possession due to the lack of [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537). See [validation](#validation) for off-chain validation.
+
+```solidity
+function registerPublicKey(bytes memory publicKey, bytes memory pop)
+```
+
+##### unregisterPublicKey
+
+Unregisters the public key associated with the given `cnNodeId` address.
+
+- The function validates the following requirements:
+- The function MUST revert if `cnNodeId` has not been registered
+- The function SHOULD authenticate and authorize `msg.sender`
+
+The function emits a `PubkeyUnregistered` event.
+
+```solidity
+function unregisterPublicKey(address cnNodeId)
+```
+
+##### getInfo
+
+Returns the public key and proof-of-possession given a `cnNodeId`.
+
+The function MUST revert if given `cnNodeId` is not registered.
+
+_Note_ The function is not able to verify the validity of the public key and the proof-of-possession due to the lack of [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537). See [validation](#validation) for off-chain validation.
+
+```solidity
+function getInfo(address cnNodeId) external view returns (BlsPublicKeyInfo pubkey)
+```
+
+##### getAllInfo
+
+Returns all the stored node IDs, public keys, and proof-of-possessions at once.
+
+_Note_ The function is not able to verify the validity of the public key and the proof-of-possession due to the lack of [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537). See [validation](#validation) for off-chain validation.
+
+```solidity
+function getAllInfo() external view returns (address[] nodeId, BlsPublicKeyInfo[])
+```
+
+### Core Logic Overview
+
+A Klaytn node must specify the target block number; starting from the target block number, the node shall read the contract and use its data. Before the target block number, the contract should be ready to use, meaning that deploying Registry contract and registering the public keys should be done.
```

Comment:
If we say that "a Klaytn node specify the target block number", this "target block number" is a configurable parameter per node? It seems like we need to define this value as a chain configuration, not a node configuration.

Reply:
No, "the target block number" indicates a predefined block number. I think it would have been "a target block number" if it was node-specific. BTW, [KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md) also mentioned it like this.

Reply:
@ian0371 Then, what about replacing "a Klaytn node" with "a Klaytn chain"?
Or, we can also rephrase the sentence like:

The target block number must be specified per chain using chainConfig;

Reply:
fixed in 1b90a54